### PR TITLE
feat: implemented `Order` and `OrderItem` entities with order endpoints

### PR DESCRIPTION
**Pull request by PavloMelnyk10**
- Implemented Order processing: placing order from user's shopping cart
- Added endpoint to view user's order history: `GET /api/orders`
- Added endpoints to retrieve order items:
  - `GET /api/orders/{orderId}/items` for all items in an order
  - `GET /api/orders/{orderId}/items/{itemId}` for a specific order item
- Implemented `PATCH /api/orders/{id}` for admins to update order status
- Added validation for request DTOs
- Created custom validator `EnumValidator` for checking `Order.Status` enum values
- Added support of pagination and soft delete
- Configured Swagger annotations for endpoint documentation

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/dto/order/UpdateOrderStatusRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import mate.academy.bookstore.model.Order;
+import mate.academy.bookstore.validation.EnumValidator;
+
+@Getter
+@Setter
+public class UpdateOrderStatusRequestDto {
+    @NotNull(message = "Status is required")
+    @EnumValidator(enumClass = Order.Status.class,
+            message = "Invalid status. Allowed values: PENDING, COMPLETED, DELIVERED")
+    private String status;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/dto/order/UpdateOrderStatusRequestDto.java:15]: you can use Order.Status

File: src/main/java/mate/academy/bookstore/mapper/OrderMapper.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package mate.academy.bookstore.mapper;
+
+import mate.academy.bookstore.config.MapperConfig;
+import mate.academy.bookstore.dto.order.CreateOrderRequestDto;
+import mate.academy.bookstore.dto.order.OrderDto;
+import mate.academy.bookstore.model.Order;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/mapper/OrderMapper.java]: should you add `uses = OrderItemMapper.class`?

File: src/main/java/mate/academy/bookstore/dto/order/OrderDto.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import lombok.Getter;
+import lombok.Setter;
+import mate.academy.bookstore.model.Order;
+
+@Getter
+@Setter
+public class OrderDto {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/dto/order/OrderDto.java:12]: your task return Order with OrderItems

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 4]
```diff
@@ -60,4 +60,21 @@ public class Book {
...
+            return false;
+        }
+        Book book = (Book) o;
+        return Objects.equals(id, book.id) && Objects.equals(isbn, book.isbn);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, isbn);
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/model/Book.java]: Why do you need it?

File: src/main/java/mate/academy/bookstore/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Status status;
+
+    @Column(nullable = false)
+    private BigDecimal total;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/model/Order.java:53]: you can use CreationTimestamp annotation

File: src/main/java/mate/academy/bookstore/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(nullable = false)
+    private boolean isDeleted;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/model/OrderItem.java]: the same

File: src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    private final UserService userService;
+    private final ShoppingCartService shoppingCartService;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Transactional
+    @Override
+    public OrderDto placeOrder(String shippingAddress) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java:42]: divide logic of this method to the private methods

File: src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.PENDING);
+
+        BigDecimal total = BigDecimal.ZERO;
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found for user: " + user.getId()));
+
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java]: can you replace it with the stream?

File: src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Override
+    @Transactional
+    public List<OrderItemDto> getOrderItems(Long orderId) {
+        User user = userService.getCurrentUser();
+
+        Order order = orderRepository.findById(orderId)
+                .filter(o -> isAdmin(user) || o.getUser().getId().equals(user.getId()))
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Order not found with id: " + orderId + " for user: " + user.getId()));
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java]: Can you filter it at the repository level?

File: src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public Page<OrderDto> getAllOrders(Pageable pageable) {
+        User user = userService.getCurrentUser();
+
+        if (isAdmin(user)) {
+            return orderRepository.findAll(pageable).map(orderMapper::toDto);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java]: It's not required in your task: GET: /api/orders (Retrieve user's order history)

File: src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                .filter(o -> isAdmin(user) || o.getUser().getId().equals(user.getId()))
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Order not found with id: " + orderId + " for user: " + user.getId()));
+
+        OrderItem orderItem = order.getOrderItems().stream()
+                .filter(item -> item.getId().equals(itemId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "OrderItem not found with id: " + itemId + " in order: " + orderId));
+
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/order/OrderServiceImpl.java]: the same. remove the admin filter and try to select the correct item on the db query level

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/mate/academy/bookstore/model/Book.java (hunk 4 above)
Comment (PavloMelnyk10) [reply src/main/java/mate/academy/bookstore/model/Book.java]: I had problems adding multiple items to a HashSet (only one order item was added), so I used these methods for testing and forgot to replace them with annotations

**Review by sarakhmen — APPROVED** [review]
lgtm1

**Pull request merged**